cmd/kaniko-acr/main.go

```diff
@@ -0,0 +1,374 @@
+}
+
+func createDockerConfig(tenantId, clientId, cert,
+	clientSecret, registry string, noPush bool) error {
+	if registry == "" {
+		return fmt.Errorf("registry must be specified")
+	}
+
+	if noPush {
+		return nil
+	}
+
+	// case of client secret or cert based auth
+	if clientId != "" {
+		// only setup auth when pushing or credentials are defined
+
+		token, err := getACRToken(tenantId, clientId, clientSecret, cert, registry)
+		if err != nil {
+			return errors.Wrap(err, "failed to fetch ACR Token")
+		}
+		err = docker.CreateDockerCfgFile(username, token, registry, dockerConfigPath)
+		if err != nil {
+			return errors.Wrap(err, "failed to create docker config")
+		}
+	} else {
+		return fmt.Errorf("managed authentication is not supported")
+	}
+
+	return nil
+}
+
+func getACRToken(tenantId, clientId, clientSecret, cert, registry string) (string, error) {
+	if tenantId == "" {
+		return "", fmt.Errorf("tenantId can't be empty for AAD authentication")
+	}
+
+	if clientId == "" {
+		return "", fmt.Errorf("clientId can't be empty for AAD authentication")
+	}
+
+	if clientSecret == "" && cert == "" {
+		return "", fmt.Errorf("one of clientSecret or cert should be defined")
+	}
+
+	// in case of authentication via cert
+	if cert != "" {
+		err := setupACRCert(cert)
+		if err != nil {
+			errors.Wrap(err, "failed to push setup cert file")
+		}
+	}
+
+	// TODO check for presence of file as well.
+	os.Setenv(clientIdEnv, clientId)
+	os.Setenv(clientSecretKeyEnv, clientSecret)
+	os.Setenv(tenantKeyEnv, tenantId)
+	env, err := azidentity.NewEnvironmentCredential(nil)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get env credentials from azure")
+	}
+
+	policy := policy.TokenRequestOptions{
+		Scopes: []string{"https://management.azure.com/.default"},
+	}
+	os.Unsetenv(clientIdEnv)
+	os.Unsetenv(clientSecretKeyEnv)
+	os.Unsetenv(tenantKeyEnv)
+	os.Unsetenv(certPathEnv)
+
+	azToken, err := env.GetToken(context.Background(), policy)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch access token")
+	}
+
+	ACRToken, err := fetchACRToken(tenantId, azToken.Token, registry)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch ACR token")
+	}
+	return ACRToken, nil
+}
+
+func fetchACRToken(tenantId, token, registry string) (string, error) {
+	formData := url.Values{
+		"grant_type":   {"access_token"},
+		"service":      {registry},
+		"tenant":       {tenantId},
+		"access_token": {token},
+	}
+	jsonResponse, err := http.PostForm(fmt.Sprintf("https://%s/oauth2/exchange", registry), formData)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch ACR token")
+	}
+	var response map[string]interface{}
+	err = json.NewDecoder(jsonResponse.Body).Decode(&response)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to decode oauth exchange response")
+	}
+
+	fmt.Printf("refrsh token " + response["refresh_token"].(string))
+	if x, found := response["refresh_token"]; found {
+		return x.(string), nil
```

Comment:
also check whether x.(string) is valid or not. x is an interface here. So, we cannot assume that it is always a string.

Reply:
done